conbench/api/compare.py

```diff
@@ -1,146 +1,273 @@
-    case_id, context_id = item["case_id"], item["context_id"]
-    key = case_id if simple else f"{case_id}-{context_id}"
-    if key not in pairs:
-        pairs[key] = {"baseline": None, "contender": None}
-    pairs[key][kind] = item
-
-
-class CompareMixin:
-    def _parse_two_ids_or_abort(self, compare_ids):
-        if "..." not in compare_ids:
-            f.abort(
-                404, description="last URL path segment must be of pattern <id>...<id>"
-            )
+    threshold = f.request.args.get("threshold")
+    if threshold is not None:
+        threshold = float(threshold)
 
-        # I think these can be either two run IDs or two
-        # benchmark result IDs?
-        baseline_id, contender_id = compare_ids.split("...", 1)
+    threshold_z = f.request.args.get("threshold_z")
+    if threshold_z is not None:
+        threshold_z = float(threshold_z)
 
-        if not baseline_id:
-            f.abort(404, description="empty baseline ID")
+    return threshold, threshold_z
 
-        if not contender_id:
-            f.abort(404, description="empty contender ID")
 
-        return baseline_id, contender_id
+def _round(value: float) -> float:
+    """Round a float to 4 significant figures, or NaN if the input is NaN."""
+    return value if math.isnan(value) else sigfig.round(value, sigfigs=4, warn=False)
 
-    def get_query_args_from_request(self):
-        """
-        Attempt to read a specific set of query parameters from request
-        context.
-        """
-        threshold = f.request.args.get("threshold")
-        if threshold is not None:
-            threshold = float(threshold)
 
-        threshold_z = f.request.args.get("threshold_z")
-        if threshold_z is not None:
-            threshold_z = float(threshold_z)
+if TYPE_CHECKING:
 
-        # I think this expression can never raise ValueError.
-        # The goal here was probably to raise ValueError?
-        # try:
-        #     baseline_id, contender_id = compare_ids.split("...", 1)
-        # except ValueError:
-        #     self.abort_404_not_found()
+    class AugmentedBenchmarkResult(BenchmarkResult):
+        """For type hints only. This represents a BenchmarkResult that has been
+        augmented in-place with the following augmentation functions, as is done in this
+        module:
 
-        return threshold, threshold_z
+        - set_z_scores()
+        - set_display_benchmark_name()
+        - set_display_case_permutation()
 
+        TODO: remove this and replace with actual BenchmarkResult properties.
+        """
 
-class CompareEntityEndpoint(ApiEndpoint, CompareMixin):
-    def _get_results(self, rid: str) -> List[BenchmarkResult]:
-        raise NotImplementedError
+        display_bmname: str
+        display_case_perm: str
+        z_score: Optional[float]
+
+
+class UnmatchingUnitsError(Exception):
+    pass
+
+
+class BenchmarkResultComparator:
+    """Dataclass to hold the comparison of two BenchmarkResults."""
```

Comment:
Thanks for this docstring, even simple things like this can make it drastically easier to read code.

Buut is it a dataclass?

Reply:
Haha, no it's not. Though it is a class meant to store data and provide an easy interface for accessing it. I can clarify ;)